features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/*
+ * Clear AES context
+ */
+void mbedtls_aes_free( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/*
+ * AES key schedule (encryption)
+ */
+int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx,
+                            const unsigned char *key,
+                            unsigned int keybits )
+{
+    if( ctx == NULL || key == NULL ) {
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+
+    if ( ( 128UL != keybits ) && ( 256UL != keybits ) ) {
+        /* Unsupported key size */
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    ctx->keybits = keybits;
+    memcpy(ctx->key, key, keybits/8);
+
+    return 0;
+}
+
+/*
+ * AES key schedule (decryption)
+ */
+int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx,
+                            const unsigned char *key,
+                            unsigned int keybits )
+{
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || key == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+
+    if ( ( 128UL != keybits ) && ( 256UL != keybits ) ) {
+        /* Unsupported key size */
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    ctx->keybits = keybits;
+
+    CRYPTO_TypeDef *device = crypto_management_acquire();
+    device->WAC = 0;
+    device->CTRL = 0;
+
+    CORE_ENTER_CRITICAL();
+    CRYPTO_KeyBufWrite(device, (uint32_t*)key, (keybits == 128) ? cryptoKey128Bits : cryptoKey256Bits);
+    CORE_EXIT_CRITICAL();
+
+    /* Busy-wait here to allow context-switching to occur */
+    device->CMD = CRYPTO_CMD_INSTR_AESENC;
+    while ((device->STATUS & CRYPTO_STATUS_INSTRRUNNING) != 0);
+
+    CORE_ENTER_CRITICAL();
+    CRYPTO_KeyRead(device, (uint32_t*)ctx->key, (keybits == 128) ? cryptoKey128Bits : cryptoKey256Bits);
+    CORE_EXIT_CRITICAL();
+
+    crypto_management_release(device);
+
+    return 0;
+}
+
+/* TODO: underneath these, we should swap out the em_crypto-provided library
+ * functions with in-place implemented functions, to get much shorter
+ * critical sections */
+
+/*
+ * AES-ECB block encryption/decryption
+ */
+int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
+                           int mode,
+                           const unsigned char input[16],
+                           unsigned char output[16] )
+{
+    int ret = 0;
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || input == NULL || output == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    if ( ctx->keybits != 128UL && ctx->keybits != 256UL) {
+        return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    CRYPTO_TypeDef *device = crypto_management_acquire();
+    device->WAC = 0;
+    device->CTRL = 0;
+
+    CORE_ENTER_CRITICAL();
```

Comment:
The resource we are protecting with critical sections is writing to or reading from 128-bit, 256-bit or 512-bit registers in our accelerator. Since that operation is done with actionable reads/writes from/to a single 32-bit register, they have to be protected against preemption.

The other registers (such as WAC and CTRL) are regular registers, which we get control over by acquiring the device. In the case of preemption, these values will be backed up and put back by the preempting code, so their use is safe.